fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -1517,6 +1518,87 @@ ACTOR Future<Void> monitorRemoteCommitted(ProxyCommitData* self) {
 	}
 }
 
+
+ACTOR Future<Void>
+proxySnapCreate(ProxySnapRequest snapReq, ProxyCommitData* commitData)
+{
+	TraceEvent("SnapMasterProxy.SnapReqEnter")
+		.detail("SnapPayload", snapReq.snapPayload)
+		.detail("SnapUID", snapReq.snapUID);
+	try {
+		// whitelist check
+		state ExecCmdValueString execArg(snapReq.snapPayload);
+		state StringRef binPath = execArg.getBinaryPath();
+		if (!isWhitelisted(commitData->whitelistedBinPathVec, binPath)) {
+			TraceEvent("SnapMasterProxy.WhiteListCheckFailed")
+				.detail("SnapPayload", snapReq.snapPayload)
+				.detail("SnapUID", snapReq.snapUID);
+			throw transaction_not_permitted();
+		}
+		// db fully recovered check
+		if (commitData->db->get().recoveryState != RecoveryState::FULLY_RECOVERED)  {
+			// Cluster is not fully recovered and needs TLogs
+			// from previous generation for full recovery.
+			// Currently, snapshot of old tlog generation is not
+			// supported and hence failing the snapshot request until
+			// cluster is fully_recovered.
+			TraceEvent("SnapMasterProxy.ClusterNotFullyRecovered")
+				.detail("SnapPayload", snapReq.snapPayload)
+				.detail("SnapUID", snapReq.snapUID);
+			throw cluster_not_fully_recovered();
+		}
+		auto result =
+			commitData->txnStateStore->readValue(LiteralStringRef("log_anti_quorum").withPrefix(configKeysPrefix)).get();
```

Comment:
Good suggestion, but there are number of LogsSets and each could be with a different tLogWriteAntiQuorum. So, the change would mean - walking the logSets and looking for locality. I will check if there is any other hint available to keep the check cheap like you mentioned, otherwise would keep it this way. Since, txnStateStore is also maintained in every proxy, probably it is comparable cost.

Reply:
I thought snapshot would fail anyways if there is more than one tlog generation? So can there be more than one tlogset?

Reply:
hmm, I suppose - because the extra tlogsets come from satellite and remote.

Reply:
ah right... I personally still prefer not to look into txnStateStore (I like stuff that provides static type safety 😉). But I guess this is a nit by now and nothing I would fight for.

Reply:
BTW: this entire block will go away because snap v2 has the ability to work with log write anti quorum configurations too, it is a matter of testing it and getting rid of it.